### PR TITLE
.github: New release workflows to simplify the developer experience.

### PR DESCRIPTION
**Pull request by cjyar**
The release process is documented in https://github.com/IronCoreLabs/ironoxide-java/blob/96-release/RELEASING.md:

1. Create and push a new branch starting with `release-v`, like `release-v0.13.0`.
1. The `release1.yaml` workflow will set the version (e.g., `0.13.0`) in various files and create a new PR for the release.
1. CI will run on the PR. Then it can either be approved and merged, or closed.
1. Merging the PR will cause `release2.yaml` to run, which builds and publishes the release.
1. After publishing the release, it increments Java-related versions to the next `-SNAPSHOT` release.

Some other points:

- CI runs on the PR before it can be approved for release.
- CI will also run on `master` after the release is made and the Java versions are incremented.
- CI now includes a job to check the version files and make sure we know how to edit them via automation.

### CODE REVIEW

File: .github/workflows/release2.yaml [hunk 1]
```diff
@@ -279,3 +313,30 @@ jobs:
...
+        keys: ${{ secrets.IRONHIDE_KEYS }}
+        input: .github/signing-key.asc.iron
+    - name: Import PGP key
+      run: gpg --batch --import .github/signing-key.asc
+    - uses: actions-rs/toolchain@v1
+      with:
+        toolchain: stable
+    - uses: actions/setup-java@v1
+      with:
+        java-version: 8
```
Comment (giarc3) [inline .github/workflows/release2.yaml:334]: Just want to double check that this is the version we want?

Comment (cjyar) [reply .github/workflows/release2.yaml:334]: @coltfred @clintfred is this correct?

File: .github/workflows/release2.yaml [hunk 2]
```diff
@@ -1,32 +1,59 @@
...
     - name: Decrypt PGP key
       uses: IronCoreLabs/ironhide-actions/decrypt@v1
       with:
         keys: ${{ secrets.IRONHIDE_KEYS }}
         input: .github/signing-key.asc.iron
     - name: Import PGP key
       run: gpg --batch --import .github/signing-key.asc
+    - name: Configure git
+      run: |
+        git config --local user.email bot@cjones.org
```
Comment (coltfred) [inline .github/workflows/release2.yaml:35]: Is this the user we want here?

File: .github/workflows/release2.yaml (hunk 1 above)
Comment (coltfred) [reply .github/workflows/release2.yaml:334]: It shouldn't matter, as we have target and source set to 8 in the build file. Maybe we should run this on 11.

https://github.com/IronCoreLabs/ironoxide-java/blob/master/java/tests/build.sbt#L25 (for reference).

Comment (coltfred) [reply .github/workflows/release2.yaml:334]: see my comment

Comment (cjyar) [reply .github/workflows/release2.yaml:334]: Okay, will switch to 11.

File: .github/workflows/release1.yaml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        echo "::set-output name=tag::${NAME}"
+        echo "::set-output name=dewey::${NUMBER}"
+        echo "::set-output name=next::${NEXT}"
+    - name: Edit files for release
+      run: .github/set-versions.sh ${{ steps.tag.outputs.dewey }}
+    - uses: actions-rs/cargo@v1
+      name: Update Cargo.lock
+      with:
+        command: update
+        args: -p ironoxide-android -p ironoxide-cpp -p ironoxide-java
```
[the remark this answers is not in the document]
Comment (cjyar) [reply .github/workflows/release1.yaml:44]: `cargo update` will update all the dependencies to the latest ones, right? Even though it's low risk, it seems like the worst possible time to change those dependencies during the process of building the release.

File: .github/workflows/release2.yaml (hunk 2 above)
Comment (cjyar) [reply .github/workflows/release2.yaml:35]: Whoops! Thanks for catching that. I did a bunch of testing in a repo on my personal account. Will fix.

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -7,6 +7,13 @@ on:
   pull_request:
 
 jobs:
+  validate:
+    runs-on: ubuntu-18.04
+    steps:
+    - uses: actions/checkout@v2
+    - name: Validate version files
+      run: .github/set-versions.sh -n 0.0.1
```
Comment (giarc3) [inline .github/workflows/ci.yaml:15]: I don't know if I understand what this is doing. `-n` makes it verify, but `0.0.1` isn't the correct version number, so it changes the versions to `0.0.1`, prints some output, then changes it back and passes the test.

Maybe this is just to make sure the single line check passes?

File: .github/workflows/release1.yaml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+# - Commit those files.
+# - Make a new PR for the release.
+# - Add a comment to the PR explaining what's going on.
+# When the release PR is merged, it triggers the second release workflow.
+
+on: create
+
+jobs:
+  kickoff:
+    if: github.event.ref_type == 'branch' && startsWith(github.event.ref, 'release-')
```
Comment (giarc3) [inline .github/workflows/release1.yaml:15]: Your comment says that the branch has to start with "release-v", but this only checks for "release-".  Which way were you intending?

File: .github/workflows/release1.yaml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        NUMBER=$(echo ${NAME} | sed -E -e 's/[^0-9.]+/./g' -e 's/\.+/./g' -e 's/^\.//' -e 's/\.$//')
+        echo "::set-output name=tag::${NAME}"
+        echo "::set-output name=dewey::${NUMBER}"
+    - name: Edit files for release
+      run: .github/set-versions.sh ${{ steps.version.outputs.dewey }}
+    - uses: actions-rs/cargo@v1
+      name: Update Cargo.lock
+      with:
+        command: update
+        args: -p ironoxide-android -p ironoxide-cpp -p ironoxide-java
```
Comment (giarc3) [inline .github/workflows/release1.yaml:44]: I'm of the opinion that we shouldn't use `actions-rs/cargo` for anything but `cross` support. I would change this section just to 
```suggestion
    - name: Update Cargo.lock
      run: cargo update -p ironoxide-android -p ironoxide-cpp -p ironoxide-java
```
Do you have a different preference?

File: .github/workflows/release2.yaml [hunk 7]
```diff
@@ -1,32 +1,59 @@
...
+# This is triggered by merging a PR from a branch named release-*. This workflow builds and publishes the release. Then it changes
+# the version to *-SNAPSHOT in various Java-related files.
 
 on:
-  push:
-    tags:
-    - '*'
+  pull_request:
+    types:
+    - closed
```
Comment (giarc3) [inline .github/workflows/release2.yaml:9]: This sounds a little scary. The comment above says the branch needs the name "release-", but where is that enforced? And does `closed` mean just merged, or if the PR is closed for any reason (perhaps that's covered by the conditional on line 13?)

File: .github/workflows/security.yaml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+name: Security Audit
```
Comment (giarc3) [inline .github/workflows/security.yaml:1]: I think that some of the things I did in [this repo](https://github.com/IronCoreLabs/unnamed-tenant-security-logging/blob/master/.github/workflows/security-audit.yaml) could be useful here as well.

File: .github/workflows/ci.yaml (hunk 4 above)
Comment (cjyar) [reply .github/workflows/ci.yaml:15]: It's verifying that it knows how to edit the version files. Making sure that nobody has moved a version file or changed its format. I'll add a comment explaining what's going on here.

File: .github/workflows/release1.yaml (hunk 5 above)
Comment (cjyar) [reply .github/workflows/release1.yaml:15]: It'll work with any branch starting with `release-`, and it'll generate a new tag with that prefix removed. Our convention for releases it that they all look like `v#.#`. But somebody could try to make a release that's not numbered, like a special release for a particular customer. Maybe that's a bad idea though; I'm not sure.

File: .github/workflows/release2.yaml (hunk 7 above)
Comment (cjyar) [reply .github/workflows/release2.yaml:9]: Thanks for catching that. I just fixed it.

File: .github/workflows/security.yaml (hunk 8 above)
Comment (cjyar) [reply .github/workflows/security.yaml:1]: I think you're right. I'll copy all that over here.

**Pull request merged**